components/blitz/src/omero/gateway/facility/DataManagerFacility.java

```diff
@@ -497,4 +572,242 @@ public void addImagesToDataset(SecurityContext ctx,
+            DSAccessException {
+        if (project != null) {
+            ProjectDatasetLink link = new ProjectDatasetLinkI();
+            link = new ProjectDatasetLinkI();
+            link.setChild(dataset.asDataset());
+            link.setParent(project.asProject());
+            link = (ProjectDatasetLink) saveAndReturnObject(ctx, link);
+            return new DatasetData(link.getChild());
+        } else {
+            return (DatasetData) saveAndReturnObject(ctx, dataset);
+        }
+    }
+    
+    /**
+     * Uploads and attaches a file to the provided {@link DataObject} (if
+     * provided)
+     * 
+     * @param ctx
+     *            {@link SecurityContext}
+     * @param file
+     *            The {@link File} to upload/attach
+     * @param mimetype
+     *            The mime type of the file (can be <code>null</code>)
+     * @param description
+     *            The description (can be <code>null</code>)
+     * @param namespace
+     *            The namespace (can be <code>null</code>)
+     * @param target
+     *            The {@link DataObject} to attach the file to (can be
+     *            <code>null</code>)
+     * @param callback
+     *            If no {@link Callback} is provided, the method will block
+     *            until the task is finished, otherwise the upload runs
+     *            asynchronously and the {@link Callback} handle gets notified
+     *            about the outcome
+     * @return The {@link FileAnnotationData} if no {@link Callback} handle was
+     *         provided, <code>null</code> otherwise
+     */
+    public FileAnnotationData attachFile(final SecurityContext ctx,
+            final File file, String mimetype, final String description,
+            final String namespace, final DataObject target, Callback callback) {
+        final String name = file.getName();
+        String absolutePath = file.getAbsolutePath();
+        final String path = absolutePath.substring(0, absolutePath.length()
+                - name.length());
+
+        final String mime;
+        if (mimetype == null)
+            mime = "application/octet-stream";
+        else
+            mime = mimetype;
+
+        final Callback cb = callback != null ? callback : new Callback();
+
+        Runnable r = new Runnable() {
+            @Override
+            public void run() {
+                RawFileStorePrx rawFileStore = null;
+                try {
+                    OriginalFile originalFile = new OriginalFileI();
+                    originalFile.setName(omero.rtypes.rstring(name));
+                    originalFile.setPath(omero.rtypes.rstring(path));
+                    originalFile.setSize(omero.rtypes.rlong(file.length()));
+                    final ChecksumAlgorithm checksumAlgorithm = new ChecksumAlgorithmI();
+                    checksumAlgorithm.setValue(omero.rtypes
+                            .rstring(ChecksumAlgorithmSHA1160.value));
```

Comment:
I don't think anybody will mind this choice but now you have me wondering if http://downloads.openmicroscopy.org/omero/5.2.0/api/slice2html/omero/grid/ManagedRepository.html#suggestChecksumAlgorithm should have been on RFS instead!